src/Mvc/Mvc.RazorPages/src/Infrastructure/CompiledPageActionDescriptorCache.cs

```diff
@@ -0,0 +1,299 @@
+        private readonly PageConventionCollection _conventions;
+        private readonly FilterCollection _globalFilters;
+        private readonly IActionDescriptorCollectionProvider _collectionProvider;
+        private readonly IPageFactoryProvider _pageFactoryProvider;
+        private readonly IPageModelFactoryProvider _modelFactoryProvider;
+        private readonly IModelBinderFactory _modelBinderFactory;
+        private readonly IRazorPageFactoryProvider _razorPageFactoryProvider;
+        private volatile InnerCache _currentCache;
+        private readonly ParameterBinder _parameterBinder;
+        private readonly IModelMetadataProvider _modelMetadataProvider;
+
+        public CompiledPageActionDescriptorCache(
+            IActionDescriptorCollectionProvider collectionProvider,
+            IEnumerable<IPageApplicationModelProvider> applicationModelProviders,
+            IViewCompilerProvider viewCompilerProvider,
+            ActionEndpointFactory endpointFactory,
+            IPageFactoryProvider pageFactoryProvider,
+            IPageModelFactoryProvider modelFactoryProvider,
+            IRazorPageFactoryProvider razorPageFactoryProvider,
+            ParameterBinder parameterBinder,
+            IModelMetadataProvider modelMetadataProvider,
+            IModelBinderFactory modelBinderFactory,
+            IOptions<RazorPagesOptions> pageOptions,
+            IOptions<MvcOptions> mvcOptions)
+        {
+            _applicationModelProviders = applicationModelProviders
+                .OrderBy(p => p.Order)
+                .ToArray();
+
+            _viewCompilerProvider = viewCompilerProvider;
+            _endpointFactory = endpointFactory;
+            _conventions = pageOptions.Value.Conventions;
+            _globalFilters = mvcOptions.Value.Filters;
+
+            _collectionProvider = collectionProvider;
+            _pageFactoryProvider = pageFactoryProvider;
+            _modelFactoryProvider = modelFactoryProvider;
+            _modelBinderFactory = modelBinderFactory;
+            _razorPageFactoryProvider = razorPageFactoryProvider;
+            _parameterBinder = parameterBinder;
+            _modelMetadataProvider = modelMetadataProvider;
+        }
+
+        private IViewCompiler Compiler => _viewCompilerProvider.GetCompiler();
+
+        private InnerCache Cache
+        {
+            get
+            {
+                var current = _currentCache;
+                var actionDescriptors = _collectionProvider.ActionDescriptors;
+
+                if (current == null || current.Version != actionDescriptors.Version)
+                {
+                    current = new InnerCache(actionDescriptors.Version);
+                    _currentCache = current;
+                }
+
+                return current;
+            }
+        }
+
+        public virtual ValueTask<CompiledPageActionDescriptor> GetOrAddAsync(PageActionDescriptor actionDescriptor)
```

Comment:
Is this really a good usage of `ValueTask`? I don't think `ValueTask` can be awaited multiple times. Also the usual pattern for this is to cache `Task<T>` instead of caching `T`